modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProcessor.java

```diff
@@ -1469,727 +1242,543 @@ private Service copyAndInject(ServiceConfiguration cfg) throws IgniteCheckedExce
+                final Map<UUID, ServiceSingleDeploymentsResults> deps = depRes.results();
 
-        if (e.getEventType() != REMOVED) {
-            svcName.set(assigns.name());
+                final HashMap<UUID, Integer> top = new HashMap<>();
+                final Collection<byte[]> errors = new ArrayList<>();
 
-            Throwable t = null;
+                deps.forEach((nodeId, res) -> {
+                    int cnt = res.count();
 
-            try {
-                redeploy(assigns);
-            }
-            catch (Error | RuntimeException th) {
-                t = th;
-            }
+                    if (cnt > 0)
+                        top.put(nodeId, cnt);
 
-            GridServiceDeploymentFuture fut = depFuts.get(assigns.name());
+                    if (!res.errors().isEmpty())
+                        errors.addAll(res.errors());
+                });
 
-            if (fut != null && fut.configuration().equalsIgnoreNodeFilter(assigns.configuration())) {
-                depFuts.remove(assigns.name(), fut);
+                if (!errors.isEmpty())
+                    fullErrors.computeIfAbsent(srvcId, e -> new ArrayList<>()).addAll(errors);
 
-                // Complete deployment futures once the assignments have been stored in cache.
-                fut.onDone(null, t);
+                fullTops.put(srvcId, top);
             }
-        }
-        // Handle undeployment.
-        else
-            undeploy(e.getKey().name());
-    }
 
-    /**
-     * @param name Name.
-     */
-    private void undeploy(String name) {
-        svcName.set(name);
+            synchronized (servicesTopsUpdateMux) {
+                updateServicesMap(registeredServices, fullTops, fullErrors);
+
+                servicesTopsUpdateMux.notifyAll();
+            }
 
-        Collection<ServiceContextImpl> ctxs;
+            ServicesExchangeActions exchangeActions = new ServicesExchangeActions();
 
-        synchronized (locSvcs) {
-            ctxs = locSvcs.remove(name);
-        }
+            exchangeActions.deploymentTopologies(fullTops);
+            exchangeActions.deploymentErrors(fullErrors);
 
-        if (ctxs != null) {
-            synchronized (ctxs) {
-                cancel(ctxs, ctxs.size());
-            }
+            msg0.servicesExchangeActions(exchangeActions);
         }
     }
 
     /**
-     *
+     * @param name Service name.
+     * @return Mapped service descriptor. Possibly {@code null} if not found.
      */
-    private static class CancelResult {
-        /** */
-        IgniteInternalFuture<?> fut;
-
-        /** */
-        boolean rollback;
-
-        /**
-         * @param fut Future.
-         * @param rollback {@code True} if service was cancelled during current call.
-         */
-        CancelResult(IgniteInternalFuture<?> fut, boolean rollback) {
-            this.fut = fut;
-            this.rollback = rollback;
+    @Nullable private ServiceInfo lookupInRegisteredServices(String name) {
```

Comment:
This is a possible point of optimization. This lookup is done in a discovery thread and may take significant time, if there are thousands of services deployed. This place should be benchmarked. I don't think, this is a blocker for the PR, but it would be good to investigate, how it affects the performance.